File: dotnet/source/examples/Publishing/TimeSeriesRangeQuery.cs
/**
 * Copyright © 2020 Push Technology Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using PushTechnology.ClientInterface.Client.Callbacks;
using PushTechnology.ClientInterface.Client.Factories;
using PushTechnology.ClientInterface.Client.Features.Control.Topics;
using PushTechnology.ClientInterface.Client.Features.TimeSeries;
using PushTechnology.ClientInterface.Client.Session;
using PushTechnology.ClientInterface.Client.Topics;
using PushTechnology.ClientInterface.Client.Topics.Details;
using static System.Console;
using static PushTechnology.ClientInterface.Examples.Runner.Program;

namespace PushTechnology.ClientInterface.Example.Publishing {
    /// <summary>
    /// Control client implementation that performs a range query of a time series.
    /// </summary>
    public sealed class PublishingTimeSeriesRangeQuery : IExample {
        /// <summary>
        /// Runs the time series range query control client example.
        /// </summary>
        /// <param name="cancellationToken">A token used to end the client example.</param>
        /// <param name="args">A single string should be used for the server url.</param>
        public async Task Run( CancellationToken cancellationToken, string[] args ) {
            string TOPIC_PREFIX = "time-series";

            var serverUrl = args[ 0 ];
            var session = Diffusion.Sessions.Principal( "control" ).Password( "password" )
                .CertificateValidation((cert, chain, errors) => CertificateValidationResult.ACCEPT)
                .Open( serverUrl );

            // Create a string topic
            var typeName = Diffusion.DataTypes.Get<String>().TypeName;
            var topicPath = $"{TOPIC_PREFIX}/{typeName}/{DateTime.Now.ToFileTimeUtc()}";
            var specification = session.TopicControl.NewSpecification(TopicType.TIME_SERIES)
                .WithProperty(TopicSpecificationProperty.TimeSeriesEventValueType, typeName);

            try
            {
                await session.TopicControl.AddTopicAsync(topicPath, specification, cancellationToken);
            }
            catch (Exception ex)
            {
                WriteLine($"Failed to add topic '{topicPath}' : {ex}.");
                session.Close();
                return;
            }

            try
            {
                await session.TimeSeries.AppendAsync<String>(topicPath, "Value1", DateTimeOffset.FromUnixTimeMilliseconds(200), cancellationToken);
                await session.TimeSeries.AppendAsync<String>(topicPath, "Value2", DateTimeOffset.FromUnixTimeMilliseconds(301), cancellationToken);
                await session.TimeSeries.AppendAsync<String>(topicPath, "Value3", DateTimeOffset.FromUnixTimeMilliseconds(301), cancellationToken);
                await session.TimeSeries.AppendAsync<String>(topicPath, "Value4", DateTimeOffset.FromUnixTimeMilliseconds(301), cancellationToken);
                await session.TimeSeries.AppendAsync<String>(topicPath, "Value5", DateTimeOffset.FromUnixTimeMilliseconds(324), cancellationToken);
                await session.TimeSeries.AppendAsync<String>(topicPath, "Value6", DateTimeOffset.FromUnixTimeMilliseconds(501), cancellationToken);
                await session.TimeSeries.AppendAsync<String>(topicPath, "Value7", DateTimeOffset.FromUnixTimeMilliseconds(501), cancellationToken);
                await session.TimeSeries.AppendAsync<String>(topicPath, "Value8", DateTimeOffset.FromUnixTimeMilliseconds(501), cancellationToken);

                await Task.Delay(TimeSpan.FromMilliseconds(300));
            }
            catch (Exception ex)
            {
                WriteLine($"Topic {topicPath} value could not be appended : {ex}.");
            }

            try
            {
                IRangeQuery<string>  rangeQuery = session.TimeSeries.RangeQuery.As<string>();

                var taskRange = rangeQuery
                    .From(DateTimeOffset.FromUnixTimeMilliseconds(301L))
                    .To(DateTimeOffset.FromUnixTimeMilliseconds(501L))
                    .SelectFromAsync(topicPath);

                taskRange.Wait(TimeSpan.FromSeconds(20));

                var results = taskRange.Result.Events.ToList();
                WriteLine($"{results.Count} results obtained from range query (301 to 501):");

                foreach(var result in results)
                {
                    WriteLine($"Sequence number: {result.Metadata.Sequence}, Value: {result.Value} with Timestamp: {result.Metadata.Timestamp}");
                }
            }
            catch (Exception ex)
            {
                WriteLine($"Range query failed for topic {topicPath} : {ex}.");
            }

            // Remove the string topic
            try
            {
                await session.TopicControl.RemoveTopicsAsync( topicPath, cancellationToken );
            } catch(Exception ex) {
                WriteLine( $"Failed to remove topic '{topicPath}' : {ex}." );
            }

            // Close the session
            session.Close();
        }
    }
}

File: python/examples/publishing/time_series_range_query.py
#   Copyright (c) 2022 Push Technology Ltd., All Rights Reserved.
#
#   Use is subject to license terms.
#
#  NOTICE: All information contained herein is, and remains the
#  property of Push Technology. The intellectual and technical
#  concepts contained herein are proprietary to Push Technology and
#  may be covered by U.S. and Foreign Patents, patents in process, and
#  are protected by trade secret or copyright law.


# Control client implementation that performs a range query of a time series.

import diffusion
from datetime import datetime, timedelta
from diffusion.features.timeseries import TimeSeries
from diffusion.datatypes import STRING

# Diffusion server connection information;
# adjust as needed for the server used in practice.
server_url = "ws://localhost:8080"
principal = "admin"
credentials = diffusion.Credentials("password")

TOPIC_PREFIX = "time-series"


async def main():
    """
    Runs the time series topic range query example.
    """
    # Creating the session.
    async with diffusion.Session(
        url=server_url, principal=principal, credentials=credentials
    ) as session:
        # Create a string topic
        data_type = STRING
        topic_path = f"{TOPIC_PREFIX}/{data_type.type_name}/{datetime.utcnow()}"
        specification = TimeSeries.of(data_type)

        try:
            await session.topics.add_topic(topic_path, specification)
        except Exception as ex:
            print(f"Failed to add topic '{topic_path}' : {ex}.")
            return

        epoch = datetime.fromtimestamp(0)
        try:
            await session.time_series.append(
                topic_path, "Value1", STRING, epoch + timedelta(milliseconds=200)
            )
            await session.time_series.append(
                topic_path, "Value2", STRING, epoch + timedelta(milliseconds=301)
            )
            await session.time_series.append(
                topic_path, "Value3", STRING, epoch + timedelta(milliseconds=301)
            )
            await session.time_series.append(
                topic_path, "Value4", STRING, epoch + timedelta(milliseconds=301)
            )
            await session.time_series.append(
                topic_path, "Value5", STRING, epoch + timedelta(milliseconds=324)
            )
            await session.time_series.append(
                topic_path, "Value6", STRING, epoch + timedelta(milliseconds=501)
            )
            await session.time_series.append(
                topic_path, "Value7", STRING, epoch + timedelta(milliseconds=501)
            )
            await session.time_series.append(
                topic_path, "Value8", STRING, epoch + timedelta(milliseconds=501)
            )

        except Exception as ex:
            print(f"Topic {topic_path} value could not be appended : {ex}.")

        try:
            range_query = session.time_series.range_query(STRING)

            range = await (
                range_query.from_(timedelta(milliseconds=301))
                .to(timedelta(milliseconds=501))
                .select_from(topic_path)
            )

            results = range.events
            print(f"{len(results)} results obtained from range query (301 to 501):")

            for result in results:
                print(
                    f"Sequence number: {result.metadata.sequence}, Value: {result.value} "
                    f"with Timestamp: {result.metadata.timestamp}"
                )
        except Exception as ex:
            print(f"Range query failed for topic {topic_path} : {ex}.")

        # Remove the string topic
        try:
            await session.topics.remove_topic(topic_path)
        except Exception as ex:
            print(f"Failed to remove topic '{topic_path}' : {ex}.")
